source/val/Instruction.h

```diff
@@ -0,0 +1,131 @@
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#ifndef LIBSPIRV_VAL_INSTRUCTION_H_
+#define LIBSPIRV_VAL_INSTRUCTION_H_
+
+#include <cstdint>
+
+#include <functional>
+#include <utility>
+#include <vector>
+
+#include "spirv-tools/libspirv.h"
+#include "table.h"
+
+namespace libspirv {
+
+class BasicBlock;
+class Function;
+
+/// Wraps the spv_parsed_instruction struct along with use and definition of the
+/// instruction's result id
+class Instruction {
+ public:
+  explicit Instruction(const spv_parsed_instruction_t* inst,
+                       Function* defining_function = nullptr,
+                       BasicBlock* defining_block = nullptr);
+
+  /// Registers the use of the Instruction in instruction \p inst at \p index
+  void RegisterUse(const Instruction* inst, uint32_t index);
+
+  uint32_t id() const { return inst_.result_id; }
+  uint32_t type_id() const { return inst_.type_id; }
+  SpvOp opcode() const { return static_cast<SpvOp>(inst_.opcode); }
+
+  /// Returns the Function where the instruction was defined. nullptr if it was
+  /// defined outside of a Function
+  const Function* function() const { return function_; }
+
+  /// Returns the BasicBlock where the instruction was defined. nullptr if it
+  /// was defined outside of a BasicBlock
+  const BasicBlock* block() const { return block_; }
+
+  /// Returns a vector of pairs of all references to this instruction's result
+  /// id. The first element is the instruction in which this result id was
+  /// referenced and the second is the index of the word in that instruction
+  /// where this result id appeared
+  const std::vector<std::pair<const Instruction*, uint32_t>>& uses() const {
+    return uses_;
+  }
+
+  /// The word used to define the Instruction
+  uint32_t words(size_t index) const { return words_[index]; }
```

Comment:
Quibble:  I'd have called this "word" instead of "words" since it only returns a single word.
Similar below with "operand(size_t)" below.
But not serious enough to block the pull request.